clang/lib/Driver/Driver.cpp

```diff
@@ -4709,6 +4733,46 @@ class OffloadingActionBuilder final {
     return C.MakeAction<OffloadAction>(HDep, DDeps);
   }
 
+  // Update Input action to reflect FPGA device archive specifics based
+  // on archive contents.
+  bool updateInputForFPGA(Action *&A, const Arg *InputArg,
+                          DerivedArgList &Args) {
+    std::string InputName = InputArg->getAsString(Args);
+    bool IsFPGAEmulation = C.getDriver().isFPGAEmulationMode();
+    // Only check for FPGA device information when using fpga SubArch.
+    if (A->getType() == types::TY_Object && isObjectFile(InputName))
+      return true;
+
+    auto ArchiveTypeMismatch = [&](bool EmitDiag) {
```

Comment:
Wonder if a micro-optimization this is possible:
```suggestion
    auto ArchiveTypeMismatch = [&C, &InputName](bool EmitDiag) {
```